DSpace/dspace-angular
# IP Authentication: X-Forwarded-For Interceptor

## Description
This PR adds an http request interceptor to the server-app module, adding the client's IP to the X-Forward-For header of each request between the node server and the REST API.
This allows IP Authentication to work properly. More details in the REST PR: [#2831](https://github.com/DSpace/DSpace/pull/2831)

## Instructions for Reviewers
Changes made:
- Implemented and configured `ForwardClientIpInterceptor` in server-app module (includes test case)

How to test:
There isn't exactly an easy way to test this, because you'd need to be able to track the requests sent between the node server and your configured REST API.
Two possible options are:
- Testing it against a REST API containing PR [#2831](https://github.com/DSpace/DSpace/pull/2831). If IP authentication works for SSR requests, the IP was correctly added to the headers.
- Adding [this interceptor](https://gist.github.com/artlowel/63b4e54febe4eef2c75308ac4dc89258) to the angular server-app module. This will log the URL and "X-Forwarded-For" header of each request sent. **Important**: Make sure the interceptor is configured **below** the `ForwardClientIpInterceptor` in the list of providers.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
